sound/soc/intel/boards/sof_rt5682.c

```diff
@@ -267,10 +268,19 @@ static int sof_rt5682_hw_params(struct snd_pcm_substream *substream,
 		}
 
 		clk_id = RT5682_PLL1_S_MCLK;
-		if (sof_rt5682_quirk & SOF_RT5682_MCLK_24MHZ)
+
+		/* get the tplg configured mclk. */
+		clk_freq = snd_sof_dai_get_ssp_mclk(rtd);
+
+		/* mclk from the quirk is the first choice */
+		if (sof_rt5682_quirk & SOF_RT5682_MCLK_24MHZ) {
+			if (clk_freq != 24000000)
+				dev_warn(rtd->dev, "configure wrong mclk in tplg, please use 24MHz.\n");
 			clk_freq = 24000000;
-		else
+		} else if (clk_freq <= 0) {
```

Comment:
`sof_rt5682_hw_params` is used for SSP only, so it is worthy to have an error message if (clk_freq < 0). And maybe we should not go further and return -EINVAL here.

Reply:
> `sof_rt5682_hw_params` is used for SSP only, so it is worthy to have an error message if (clk_freq < 0). And maybe we should not go further and return -EINVAL here.

Do we really want to return fail? I mean we don't have reading from tplg with today's code, and we use default 19.2MHz there, do we really want to change the behavior?

Reply:
> > `sof_rt5682_hw_params` is used for SSP only, so it is worthy to have an error message if (clk_freq < 0). And maybe we should not go further and return -EINVAL here.
> 
> Do we really want to return fail? I mean we don't have reading from tplg with today's code, and we use default 19.2MHz there, do we really want to change the behavior?

`clk_freq < 0` will only happen if `dai->dai_config->type` is not SOF_DAI_INTEL_SSP. IMHO, we should never meet the `clk_freq < 0` case here and there should something wrong if we see it. That's why I suggest we should return error here. And I don't think the behavior will be changed since we should not meet the `clk_freq < 0` case.

Reply:
basically, its 'int' from topology, so it could be configured to be minus value by mistake also. By the way, I feel it is easy to be convoluted with this kind of github comments, @bardliao can you help to post what code change here you suggested directly?

Reply:
@keyonjie See below.
```
		/* mclk from the quirk is the first choice */
		if (sof_rt5682_quirk & SOF_RT5682_MCLK_24MHZ) {
			if (clk_freq != 24000000)
				dev_warn(rtd->dev, "configure wrong mclk in tplg, please use 24MHz.\n");
			clk_freq = 24000000;
		} else if (clk_freq == 0) {
			/* use default mclk if not specified correct in topology */
			clk_freq = 19200000;
		} else if (clk_freq < 0) {
			dev_err(rtd->dev, "dai_config->type should be SSP err: %d\n", clk_freq);
			return clk_freq;
		}
```

Reply:
@bardliao thanks, got your point now. 

Looks legit, but let me move the 'dev_err' logging to the snd_sof_dai_get_ssp_mclk() helper which could be more easier for understanding.